src/CacheManager.php

```diff
@@ -10,6 +10,8 @@
 
 class CacheManager extends BaseCacheManager
 {
+    public static bool $addTags = true;
```

Comment:
Wouldn't the fact that:
1. The CacheManager is now always replaced in the TenancyServiceProvider
2. This defaults to `true`

Lead to this feature being always enabled by default? What is the purpose of `CacheTenancyBootstrapper` now and how does it being enabled/disabled affect the actual behavior?

Reply:
> What is the purpose of CacheTenancyBootstrapper now and how does it being enabled/disabled affect the actual behavior?

Since the CacheManager is replaced in the SP, the bootstrapper seems redundant. Using it together with PrefixCacheTenancyBootstrapper doesn't seem to affect the behavior in any way.

The feature (cache prefixing) is not enabled by default – because the `$addTags` property is `true`, the default behavior of the cache manager is basically the same as it was before, but now, we replace the CacheManager instance in the SP instead of the bootstrapper.

Caching should work the same way as it worked before (except the need for using CacheTenancyBootstrapper) – the cache gets scoped by the tags (only when tenancy is initialized), and to change that behavior so that prefixing is used instead of tagging, we'd have to 1) change `$addTags` to `false` and 2) use PrefixCacheTenancyBootstrapper.

Reply:
I just deleted CacheTenancyBootstrapper, and everything seems to be working fine

Reply:
As I wrote in a review just now, we don't want to remove it. We need backwards compatibility. And we still want to offer both approaches to separating tenant caches.

Reply:
> we still want to offer both approaches to separating tenant caches

There's probably a misunderstanding – as I wrote, both approaches are still supported, it's just that CacheTenancyBootstrapper is now useless because its point was replacing the cache manager, and that's done in the SP now ("Caching should work the same way as it worked before (except the need for using CacheTenancyBootstrapper) – the cache gets scoped by the tags (only when tenancy is initialized), and to change that behavior so that prefixing is used instead of tagging, we'd have to 1) change $addTags to false and 2) use PrefixCacheTenancyBootstrapper.").

> we don't want to remove it. We need backwards compatibility

We're replacing the cache manager in the SP, so I don't think we can make this completely backwards compatible (removed cache bootstrapper can exist, but its only purpose would be just that – to be there so that users won't be getting an exception).

So backwards compatibility seems like the only reason to keep the bootstrapper. But couldn't we mention that the bootstrapper got removed and that it should be removed from the `tenancy.bootstrappers` config, since now, Tenancy replaces the cache manager in the SP?

Reply:
I see, so controlling whether tags are used is now done purely using the static property?

If so, maybe the bootstrapper could be used for that? It'd feel kind of in line with how other types of "tenancy logic" are enabled by the user. Also, enabling cache tags by default **without the user seeing it anywhere** (as they do with bootstrappers — they can see which ones are not commented) doesn't seem ideal.

Also, what again was the reason for moving the logic to the SP?

> So backwards compatibility seems like the only reason to keep the bootstrapper

When I refer to backwards compatibility I only mean the behavior. Syntax can change in any way 👍🏻

Reply:
> what again was the reason for moving the logic to the SP?

If the logic was in a bootstrapper, changing the cache manager instance scopes the cache to the current tenant completely.

E.g., do `cache(['key' => 'central'])` in the central context, switch to the tenant context, do `cache('key')`. If the prefix bootstrapper is used and it changes the cache manager instead of doing that in the SP, we'd expect the unprefixed stores to return 'central'. But because we're swapping the cache manager entirely, we'd get `null`.

I think for prefixing, we can't swap the cache manager instances in the bootstrappers. We have to do it only once at some point (TSP) so that the cache only gets prefixed (`cache('key')` => `cache('tenant.key')`), not _fully_ scoped if that makes sense. And if we're using the prefix bootstrapper, but we're using a cache store that we don't want to prefix, we expect the cache calls to work as if they were central (unprefixed, not scoped), so we can't really bind a new CacheManager instance in the bootstrapper.

Reply:
> I see, so controlling whether tags are used is now done purely using the static property?
>
> If so, maybe the bootstrapper could be used for that?

Added a bootstrapper for enabling the tags here https://github.com/archtechx/tenancy/pull/1014/commits/7ebfc375d6ef0f43311544b26750020b9f905a6f